File: section2/101.对称二叉树.py
#
# @lc app=leetcode.cn id=101 lang=python3
#
# [101] 对称二叉树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __str__(self):
        return f'node(val={self.val})'

    @staticmethod
    def generate(nums: list):
        root = TreeNode(nums[0])
        nodes = [root]

        index = 1
        while index < len(nums):
            new_nodes = []
            for node in nodes:
                if node:
                    node.left = None if nums[index] is None else TreeNode(nums[index])
                    node.right = None if nums[index + 1] is None else TreeNode(nums[index + 1])

                new_nodes.append(node.left if node else None)
                new_nodes.append(node.right if node else None)
                index += 2

            nodes = new_nodes

        return root


class Solution:
    def is_same_by_recursive(self, left: TreeNode, right: TreeNode) -> bool:
        if not left and not right:
            return True

        if left and right and left.val == right.val:
            # 对于每个对称节点判断他俩左右节点的值
            return self.is_same_by_recursive(left.left, right.right) and self.is_same_by_recursive(left.right, right.left)

        return False

    def is_same(self, left: TreeNode, right: TreeNode) -> bool:
        """
        使用栈的方式来模拟递归算法
        """
        stacks = [(left, right)]
        while stacks:
            left_node, right_node = stacks.pop()
            if left_node and right_node and left_node.val == right_node.val:
                stacks.append((left_node.left, right_node.right))
                stacks.append((left_node.right, right_node.left))
            elif left_node or right_node:
                return False

        return True

    def isSymmetric(self, root: TreeNode) -> bool:
        return self.is_same(root.left, root.right)

    def test(self):
        print(self.isSymmetric(TreeNode.generate([1, 2, 2, 3, 4, 4, 3])))
        print(not self.isSymmetric(TreeNode.generate([1, 2, 2, None, 3, None, 3])))


Solution().test()

# @lc code=end

File: section2/section/101.对称二叉树.go
package section

import "fmt"

/*
 * @lc app=leetcode.cn id=101 lang=golang
 *
 * [101] 对称二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isEqual(left, right *TreeNode) bool {
	if left == nil && right == nil {
		return true
	}

	if left == nil || right == nil {
		return false
	}

	return left.Val == right.Val && isEqual(left.Left, right.Right) && isEqual(left.Right, right.Left)
}

//使用迭代方式来进行镜像校验
func isSymmetricByRange(root *TreeNode) bool {
	currents := []*TreeNode{root}
	for len(currents) > 0 {
		for i := 0; i < (len(currents) / 2); i++ {
			compare := currents[len(currents)-i-1]
			if currents[i] != nil && compare != nil {
				if currents[i].Val != compare.Val {
					return false
				}
			} else {
				if currents[i] != compare {
					return false
				}
			}
		}

		nextCurrents := make([]*TreeNode, 0, len(currents)*2)
		for _, current := range currents {
			if current != nil {
				nextCurrents = append(nextCurrents, current.Left, current.Right)
			}
		}

		currents = nextCurrents
	}

	return true
}

func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}

	return isEqual(root.Left, root.Right)
}

func Test101() {
	fmt.Println(isSymmetric(ToNode([]int{1, 2, 2, 3, 4, 4, 3})))
	fmt.Println(!isSymmetric(ToNode([]int{1, 2, 2, 0, 3, 0, 3})))
}

// @lc code=end
